Lib/unittest/loader.py

```diff
@@ -283,19 +283,29 @@ def discover(self, start_dir, pattern='test*.py', top_level_dir=None):
             sys.path.insert(0, top_level_dir)
         self._top_level_dir = top_level_dir
 
-        is_not_importable = False
         is_namespace = False
         tests = []
         if os.path.isdir(os.path.abspath(start_dir)):
-            start_dir = os.path.abspath(start_dir)
-            if start_dir != top_level_dir:
-                is_not_importable = not os.path.isfile(os.path.join(start_dir, '__init__.py'))
+            if os.path.abspath(start_dir) != top_level_dir:
+                try:
+                    # Convert to dot import
+                    dot_import = '.'.join(
+                        os.path.relpath(start_dir, top_level_dir).split(os.sep)
+                    )
+                    __import__(dot_import)
+                except ImportError:
+                    raise ImportError('Start directory is not importable: %r' % start_dir)
+
+            if not os.path.isfile(os.path.join(start_dir, '__init__.py')):
+                is_namespace = True
+
+            tests = list(self._find_tests(start_dir, pattern, is_namespace))
         else:
             # support for discovery from dotted module names
             try:
                 __import__(start_dir)
             except ImportError:
-                is_not_importable = True
+                raise ImportError('Start directory is not importable: %r' % start_dir)
```

Comment:
I don't know if is convenient raise a Exception here. This can break other tests